rust-runtime/aws-smithy-mocks-experimental/src/lib.rs

```diff
@@ -0,0 +1,220 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/* Automatically managed default lints */
+#![cfg_attr(docsrs, feature(doc_auto_cfg))]
+/* End of automatically managed default lints */
+use std::fmt::{Debug, Formatter};
+use std::future::Future;
+use std::marker::PhantomData;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::Arc;
+
+use aws_smithy_runtime_api::box_error::BoxError;
+use aws_smithy_runtime_api::client::interceptors::context::{
+    BeforeDeserializationInterceptorContextMut, BeforeSerializationInterceptorContextMut, Error,
+    FinalizerInterceptorContextMut, Input, Output,
+};
+use aws_smithy_runtime_api::client::interceptors::Intercept;
+use aws_smithy_runtime_api::client::orchestrator::{HttpResponse, OrchestratorError};
+use aws_smithy_runtime_api::client::result::SdkError;
+use aws_smithy_runtime_api::client::runtime_components::RuntimeComponents;
+use aws_smithy_runtime_api::http::{Response, StatusCode};
+use aws_smithy_types::body::SdkBody;
+use aws_smithy_types::config_bag::{ConfigBag, Storable, StoreReplace};
+
+#[macro_export]
+macro_rules! mock {
+    ($operation: expr) => {
+        #[allow(unreachable_code)]
+        {
+            $crate::RuleBuilder::new(
+                || $operation(todo!()).as_input().clone().build().unwrap(),
+                || $operation(todo!()).send(),
+            )
+        }
+    };
+}
+
+type MatchFn = Arc<dyn Fn(&Input) -> bool + Send + Sync>;
+type OutputFn = Arc<dyn Fn() -> Result<Output, OrchestratorError<Error>> + Send + Sync>;
+
+impl Debug for MockResponseInterceptor {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{} functions", self.rules.len())
+    }
+}
+
+#[derive(Clone)]
+pub enum MockOutput {
+    HttpResponse(Arc<dyn Fn() -> Result<HttpResponse, BoxError> + Send + Sync>),
+    ModeledResponse(OutputFn),
+}
+
+/// Interceptor which produces mock responses based on a list of rules
+pub struct MockResponseInterceptor {
+    rules: Vec<Rule>,
+}
+
+pub struct RuleBuilder<I, O, E> {
+    _ty: PhantomData<(I, O, E)>,
+    input_filter: MatchFn,
+}
+
+impl<I, O, E> RuleBuilder<I, O, E>
+where
+    I: Send + Sync + Debug + 'static,
+    O: Send + Sync + Debug + 'static,
+    E: Send + Sync + Debug + std::error::Error + 'static,
+{
+    pub fn new<F, R>(_input_hint: impl Fn() -> I, _output_hint: impl Fn() -> F) -> Self
+    where
+        F: Future<Output = Result<O, SdkError<E, R>>>,
+    {
+        Self {
+            _ty: Default::default(),
+            input_filter: Arc::new(|i: &Input| i.downcast_ref::<I>().is_some()),
+        }
+    }
+
+    /// Add an additional filter to constrain which inputs match this rule
+    pub fn match_requests(mut self, filter: impl Fn(&I) -> bool + Send + Sync + 'static) -> Self {
```

Comment:
Sometimes request inputs are going to be identical. There needs to be a way to differentiate them. Maybe add a public CallIndex to the config bag?